EIP-5218: NFT Rights Management (#5222)
ethereum/EIPs: EIPS/eip-5218.md

---

> eip: 5218

Reviewer: this pr has number 5222

Reviewer 2: I assigned 5218 because 5222 seems a bit too "nice" a number, and since you cannot prove/disprove number squatting, I assigned an unused number.

Author: @TimDaub Shall I resolve this conversation?

Reviewer 3: > I assigned 5218 because 5222 seems a bit too "nice" a number, and since you cannot prove/disprove number squatting, I assigned an unused number.

Usually the burden of proof is on the accuser, why is it different here?

Reviewer 4: EIP numbers are assigned by the editors.  It is intentionally left up to the editors to decide what number to give so that people cannot game the system.  While we do have a default rule for number selection that we use internally to simplify our lives, we often override that rule in order to defend against number sniping because in the past number sniping has lead to people spamming the repository with garbage issues in order to farm for certain numbers.

There is no "proof" necessary for anything here, the editor assigned a number is all that has happened here.

Reviewer 3: The editor assigned a number because the initial one was "too nice" according to their own words, please see comments above. There is zero indication that the authors of this EIP spammed to end up with the number they initially did. So how exactly is the editor preventing spam here?

Reviewer 2: I'll resolve this.
```suggestion
eip: 5222
```

Reviewer 4: Editors assign numbers, not authors and not the PR number (even though that is what we *sometimes* use internally to simplify assignment).  There is nothing wrong with 5218 as an EIP number, and I see no reason to change it to 5222 in this case.

Reviewer 4: It isn't possible to prove someone is gaming the system because it is a very fuzzy line, and we have a history of problems where people spam create issues on this repository to bump the number so they can snipe (which creates work for the editors who are woefully understaffed).  It is far easier to just have the simple rule of "editor picks the number" (which is what we have).

The editors try to give their reasoning for not following our usual **internal** number selection mechanism to facilitate understanding between and potentially start a conversation among editors.  Sometimes disagreements occur between editors and they are usually discussed in our fortnightly meetings.

What is pertinent about the above is that editors choose the EIP number, not authors, and there is no mechanism currently for authors to choose their preferred EIP numbers.  We have an internal mechanism for number selection that we usually follow because it is easy, but we don't stick strictly to it specifically so we can address any potential gaming of the system.

Since numbers are just a UID, we try to be over-cautious since false positives have no real cost.  This frees us from having to waste too much time thinking about system gaming as we can just freely assign numbers as we deem necessary, even if it turns out it isn't necessary for some reason.

----

It is worth noting that editors are all volunteers and no one is getting paid for their work.  We are woefully understaffed and overworked so most of our rules are around reducing editor workload as much as possible while maintaining as much fairness as we can.  I personally prioritize helping people understand the EIP process whenever I can, but it is one of the major editor time sinks.  Arguing with the editors over their decisions is more likely to cause the rules to become even more draconian, not less, as we will need to adjust them to be even stricter if enough people cause editors to spend time explaining why we reserve the right to pick numbers or external links are not allowed.

Anyone is welcome to participate in the EIP improvement process and/or EIPIP meetings where we discuss these things, but be forewarned that any solution that increases editor time or decreases fairness (even transitively) is unlikely to see much editor support.

title: NFT Rights Management Standard
description: An interface for creating copyright licenses that transfer with an NFT.
status: Draft
type: Standards Track
category: ERC

> author: James Grimmelmann <james.grimmelmann@cornell.edu>, Yan Ji <jyamy42@gmail.com>, Tyler Kell <sk3259@cornell.edu>

Reviewer 4: At least one author must have a GitHub handle.

Author: Github handles added. Thank you!

---
eip: 5218
title: NFT Rights Management Standard
description: An interface for creating copyright licenses that transfer with an NFT.
status: Draft
type: Standards Track
category: ERC
author: James Grimmelmann <james.grimmelmann@cornell.edu>, Yan Ji <jyamy42@gmail.com>, Tyler Kell <sk3259@cornell.edu>

> discussions-to: https://github.com/ethereum/eips/issues/5218

Reviewer 2: Must be a thread on ethereum magicians

Author: Thank you for editing the urls! I've created a thread on ethereum magicians and updated the link.

> created: 2022-7-11

Reviewer 2: ```suggestion
created: 2022-07-11
```
Minor formatting error

> requires: 721

Reviewer 2: ```suggestion
requires: 721, 5218
```
References 5218

---

> ## Table of Contents ##
> 
> * [Simple Summary](#simple-summary)
> * [Abstract](#abstract)
> * [Motivation](#motivation)
> * [Specification](#specification)
> * [Rationale](#rationale)
> * [Backwards Compatibility](#backwards-compatibility)
> * [Test Cases](#test-cases)
> * [Reference Implementation](#reference-implementation)
> * [Security Considerations](#security-considerations)
> * [Copyright](#copyright)
> * [Citation](#citation)

Reviewer 2: ```suggestion
```
No longer a valid EIP section (generated by renderer)

> ## Simple Summary
> 
> A standard interface for creating copyright licenses that transfer with an NFT.

Reviewer 2: ```suggestion
```

## Abstract

The following standard defines an API for managing NFT licenses. This standard provides basic functionality to create, transfer, and revoke licenses, and to determine the current licensing state of an NFT. The standard does not define the legal details of the license. Instead, it provides a structured framework for recording licensing details.

We consider use cases of NFT creators who wish to give the NFT holder a copyright license to use a work associated with the NFT. The license can optionally be revoked under conditions specified by the creator. The holder of an active license can issue sublicenses to others to carry out the rights granted under the license.

## Motivation

> The [ERC-721](https://eips.ethereum.org/EIPS/eip-721) standard defines an API to track and transfer ownership of an NFT. When an NFT is to represent some off-chain asset, however, we would need some legally effective mechanism to *tether* the on-chain asset (NFT) to the off-chain property. One important case of off-chain property is a creative work such as an image or music file. Recently, most NFT projects involving creative works have used licenses to clarify what legal rights are granted to the NFT owner. But these licenses are almost always off-chain and the NFTs themselves do not indicate what licenses apply to them, leading to uncertainty about rights to use the work associated with the NFT. It is not a trivial task to avoid all the [copyright vulnerabilities in NFTs](https://medium.com/initc3org/copyright-vulnerabilities-in-nfts-317e02d8ae26), nor have existing EIPs addressed rights management of NFTs beyond the simple cases of [direct ownership](https://eips.ethereum.org/EIPS/eip-721) or [rental](https://eips.ethereum.org/EIPS/eip-4907).

Reviewer 2: ```suggestion
The [EIP-721](./eip-721.md) standard defines an API to track and transfer ownership of an NFT. When an NFT is to represent some off-chain asset, however, we would need some legally effective mechanism to *tether* the on-chain asset (NFT) to the off-chain property. One important case of off-chain property is creative work such as an image or music file. Recently, most NFT projects involving creative works have used licenses to clarify what legal rights are granted to the NFT owner. But these licenses are almost always off-chain and the NFTs themselves do not indicate what licenses apply to them, leading to uncertainty about rights to use the work associated with the NFT. It is not a trivial task to avoid all the copyright vulnerabilities in NFTs, nor have existing EIPs addressed rights management of NFTs beyond the simple cases of direct ownership (see [EIP-721](./eip-721.md)) or rental (see [EIP-4907](./eip-4907.md)).
```

> This EIP attempts to provide a standard to facilitate rights management of NFTs in the world of Web3. In particular, ERC-5218 smart contracts allow all licenses to an NFT, including the *root license* issued to the NFT owner and *sublicenses* granted by a license holder, to be recorded and easily tracked with on-chain data. These licenses can consist of human-readable legal code, machine-readable summaries such as those written in [CC REL](https://wiki.creativecommons.org/wiki/CC_REL), or both. An EIP-5218 smart contract points to a license by recording a URI, providing a reliable reference for users to learn what legal rights they are granted and for NFT creators and auditors to detect unauthorized infringing uses.

Reviewer 2: ```suggestion
This EIP attempts to provide a standard to facilitate rights management of NFTs in the world of Web3. In particular, ERC-5218 smart contracts allow all licenses to an NFT, including the *root license* issued to the NFT owner and *sublicenses* granted by a license holder, to be recorded and easily tracked with on-chain data. These licenses can consist of human-readable legal code, machine-readable summaries such as those written in CC REL, or both. An EIP-5218 smart contract points to a license by recording a URI, providing a reliable reference for users to learn what legal rights they are granted and for NFT creators and auditors to detect unauthorized infringing uses.
```
No external links

## Specification

> The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Reviewer 2: ```suggestion
The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
```

The full log of license creation, transfer, and revocation *must* be traceable via event logs. Therefore, all license creations and transfers *must* emit a corresponding log event. Revocation may differ a bit. An implementation of this EIP may emit a `Revoke` event only when a license is revoked in a function call, or for every revoked license, both are sufficient to trace the status of all licenses. The former costs less gas if revoking a license automatically revokes all sublicenses under it, while the latter is efficient in terms of interrogation of a license status. Implementers should make the tradeoffs depending on their license terms.

The `revoker` of a license may be the licensor, the license holder, or a smart contract address which calls the `revokeLicense` function when some conditions are met. Implementers should be careful with the authorization, and may make the `revoker` smart contract forward compatible with transfers by not hardcoding the addresses of `licensor` or `licenseHolder`.

> The license `URI` may point to a JSON file that conforms to the "ERC5218 Metadata JSON Schema" as below, which adopts the ["three-layer"](https://creativecommons.org/licenses/) design of the Creative Commons Licenses:

Reviewer 2: ```suggestion
The license `URI` may point to a JSON file that conforms to the "ERC5218 Metadata JSON Schema" as below, which adopts the "three-layer" design of the Creative Commons Licenses:
```

```json
{
    "title": "License Metadata",
    "type": "object",
    "properties": {
        "legal-code": {
            "type": "string",
            "description": "The legal code of the license."
        },
        "human-readable": {
            "type": "string",
            "description": "The human readable license deed."
        },
        "machine-readable": {
            "type": "string",
            "description": "The machine readable code of the license that can be recognized by software, such as CC REL."
        }
    }
}
```

> The `supportsInterface` method MUST return `true` when called with `0xXXX(TODO)`.

Reviewer 2: The TODO is a blocker for merging

## Rationale

> This EIP aims to allow tracing all licenses to an NFT to facilitate right management. The ERC-721 standard only logs the property but not the legal rights tethered to NFTs. Even when logging the license via the optional ERC721Metadata extension, sublicenses are not traceable, which doesn't comply with the transparency goals of Web3. Some implementations attempt to get around this limitation by minting NFTs to represent a particular license, such as the [BAYC #6068 Royalty-Free Usage License](https://opensea.io/assets/0x495f947276749ce646f68ac8c248420045cb7b5e/111558001269395431138572590975405607139590663093731712964826342979501444572944). This is not an ideal solution because the linking between different licenses to an NFT is ambiguous. An auditor has to investigate all NFTs in the blockchain and inspect the metadata which hasn't been standardized in terms of sublicense relationship. To avoid these problems, this EIP logs all licenses to an NFT in a tree data structure, which is compatible with ERC721 and allows efficient traceability.

Reviewer 2: ```suggestion
This EIP aims to allow tracing all licenses to an NFT to facilitate right management. The ERC-721 standard only logs the property but not the legal rights tethered to NFTs. Even when logging the license via the optional ERC721Metadata extension, sublicenses are not traceable, which doesn't comply with the transparency goals of Web3. Some implementations attempt to get around this limitation by minting NFTs to represent a particular license, such as the BAYC #6068 Royalty-Free Usage License. This is not an ideal solution because the linking between different licenses to an NFT is ambiguous. An auditor has to investigate all NFTs in the blockchain and inspect the metadata which hasn't been standardized in terms of sublicense relationship. To avoid these problems, this EIP logs all licenses to an NFT in a tree data structure, which is compatible with ERC721 and allows efficient traceability.
```

## Backwards Compatibility

This standard is compatible with the current ERC-721 standards: a contract can inherit from both ERC-721 and EIP-5218 at the same time.

## Test Cases

> Test cases are available [here](https://eips.ethereum.org/assets/eip-5218/test/Contract.t.sol).

Reviewer 2: ```suggestion
Test cases are available [here](../assets/eip-5218/test/Contract.t.sol).
```

```

Each NFT has a license tree and starting from each license, one can trace back to the root license via `parentLicenseId` along the path. 

In the reference implementation, once a license is revoked, all sublicenses under it are revoked. This is realized in a *lazy* manner for lower gas cost, i.e., assign `active=false` only for licenses that are explicitly revoked in a `revokeLicense` function call. Therefore, `isLicenseActive` returns `true` only if all its ancestral licenses haven't been revoked.

For non-root licenses, the creation, transfer and revocation are straightforward:

1. Only the holder of an active license can create sublicenses.
2. Only the holder of an active license can transfer it to a different license holder. 
3. Only the revoker of an active license can revoke it.

The root license must be compatible with `ERC-721`: 

1. When an NFT is minted, a license is granted to the NFT owner.
2. When an NFT is transferred, the license holder is changed to the new owner of the NFT.
3. When a root license is revoked, the NFT is returned to the NFT creator, and the NFT creator may later transfer it to a new owner with a new license.

> The complete implementation can be found [here](https://eips.ethereum.org/assets/eip-5218/contracts/RightsManagement.sol) and reference license terms [here](TODO).

Reviewer 2: ```suggestion
The complete implementation can be found [here](../assets/eip-5218/contracts/RightsManagement.sol) and reference license terms [here](TODO).
```
The TODO is a blocker for merging

Author: Thank you for noting that. The TODOs are addressed.

## Security Considerations

Implementors of the `IERC5218` standard must consider thoroughly the permissions they give to `licenseHolder` and `revoker`. If the license is ever to be transferred to a different license holder, the `revoker` smart contract should not hardcode the `licenseHolder` address to avoid undesirable scenarios.

## Copyright

> Copyright and related rights waived via [CC0](https://eips.ethereum.org/LICENSE).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

> ## Citation
> 
> Please cite this document as:
> 
> [James Grimmelmann](james.grimmelmann@cornell.edu), [Yan Ji](jyamy42@gmail.com), [Tyler Kell](sk3259@cornell.edu), "EIP-5218: NFT Rights Management Standard," *Ethereum Improvement Proposals*, no. 5218, June 2022. [Online serial]. Available: https://eips.ethereum.org/EIPS/eip-5218.

Reviewer 2: ```suggestion
```

No longer a valid EIP section (generated by renderer)
